### PR TITLE
Add jukebox

### PR DESCRIPTION
**Pull request by alexey-petrov**
Closes https://github.com/sparkletown/internal-sparkle-issues/issues/1071
follow up for https://github.com/sparkletown/sparkle/pull/2209
(re created for staging branch)

### CODE REVIEW

File: src/types/jukebox.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import firebase from "firebase/app";
+
+export type JukeboxMessage = {
+  from: string;
+  text: string;
+  tableId: string;
+  ts_utc: firebase.firestore.Timestamp;
```
Comment (sunny-viktoryia) [inline src/types/jukebox.ts:7]: have you found a way to parse the firestore timestamp? I couldn't do that. 🙈 

if you know, please, share how can we do that.
If you don't, then let's stick to `utcMilliseconds` -> `Date.now()` to be able to display the date/time when needed.

File: src/types/venues.ts [hunk 2]
```diff
@@ -181,6 +181,7 @@ export interface BaseVenue {
   showRangers?: boolean;
   chatTitle?: string;
   showReactions?: boolean;
+  enableJukebox?: boolean;
```
Comment (sunny-viktoryia) [inline src/types/venues.ts:184]: can we extend `JazzBarVenue` (not `BaseVenue`) instead?

File: src/utils/jukebox.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import firebase from "firebase/app";
+
+import { JukeboxMessage } from "types/jukebox";
+
+export const buildJukeboxMessage = <T extends JukeboxMessage>(
+  message: Pick<T, Exclude<keyof T, "ts_utc">>
+) => ({
+  ...message,
+  ts_utc: firebase.firestore.Timestamp.now(),
```
Comment (sunny-viktoryia) [inline src/utils/jukebox.ts:9]: a question about Firestore timestamp. (check another comment)

File: src/utils/url.ts [hunk 4]
```diff
@@ -105,6 +105,21 @@ export const isValidUrl = (url: string): boolean => {
   }
 };
 
+export const isStringAValidUrl = (urlString: string) => {
```
Comment (sunny-viktoryia) [inline src/utils/url.ts:108]: why have you created a new function while there is `isValidUrl`? can you refactor the existing one?

File: src/components/molecules/Jukebox/Jukebox.scss [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    place-content: center;
+    flex-grow: 1;
+    margin-right: $spacing--lg;
+  }
+
+  &__input-container {
+    flex-grow: 1;
+    display: flex;
+
+    > div {
```
Comment (sunny-viktoryia) [inline src/components/molecules/Jukebox/Jukebox.scss:50]: let's give a class name to this div if possible. 🙃

File: src/components/molecules/Jukebox/Jukebox.scss [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    padding-right: $submit-button-diameter;
+    // NOTE: this removes screen artefacts of the embedded <input> inside InputField being as wide as the parent
+    height: $spacing--lg * 2;
+  }
+
+  &__buttons {
+    display: flex;
+    flex-grow: 0;
+
+    & .ButtonNG__button {
```
Comment (sunny-viktoryia) [inline src/components/molecules/Jukebox/Jukebox.scss:66]: can we move it out of `&__buttons` not to increase the specificity of the css rulesets?

File: src/components/templates/Jazzbar/JazzTab/JazzTab.tsx [hunk 7]
```diff
@@ -121,13 +129,27 @@ const Jazz: React.FC<JazzProps> = ({ setUserList, venue }) => {
   const shouldShowReactions =
     (seatedAtTable && venue.showReactions) ?? DEFAULT_SHOW_REACTIONS;
 
+  const shouldShowJukebox =
+    (!!seatedAtTable &&
+      venue.enableJukebox &&
+      seatedAtTable === MUSIC_VENUE_FIRST_TABLE_NAME) ??
```
Comment (sunny-viktoryia) [inline src/components/templates/Jazzbar/JazzTab/JazzTab.tsx:135]: why only first table?

File: src/components/templates/Jazzbar/JazzTab/JazzTab.tsx [hunk 8]
```diff
@@ -167,26 +189,28 @@ const Jazz: React.FC<JazzProps> = ({ setUserList, venue }) => {
...
-                    frameBorder="0"
-                    allow={IFRAME_ALLOW}
-                  />
-                ) : (
-                  <div className="iframe-video">
-                    Embedded Video URL not yet set up
-                  </div>
-                )}
-              </div>
+              {
```
Comment (sunny-viktoryia) [inline src/components/templates/Jazzbar/JazzTab/JazzTab.tsx:196]: why was that added? is anything changed here?

File: src/components/templates/Jazzbar/JazzTab/JazzTab.tsx [hunk 9]
```diff
@@ -225,6 +258,7 @@ const Jazz: React.FC<JazzProps> = ({ setUserList, venue }) => {
             setUserList={setUserList}
             setSeatedAtTable={setSeatedAtTable}
             isAudioEffectDisabled={isUserAudioMuted}
+            shouldShowJukebox={shouldShowJukebox}
```
Comment (sunny-viktoryia) [inline src/components/templates/Jazzbar/JazzTab/JazzTab.tsx:261]: why is that needed here?

File: src/utils/url.ts [hunk 10]
```diff
@@ -105,6 +105,21 @@ export const isValidUrl = (url: string): boolean => {
...
+
+    return VALID_URL_PROTOCOLS.includes(url.protocol);
+  } catch (e) {
+    if (e instanceof Error) {
+      return false;
+    }
+    throw e;
+  }
+};
+
```
Comment (dmitrymarchuk) [inline src/utils/url.ts:122]: This is identical to the function `isValidUrl` above except for the first line. And for the `instanceof` check.
As this is used in one place you can simply check `if (!urlString) return false;` for that in that one place and use `isValidUrl` there. Also please replace `e.name === "TypeError" with `instanceof` check inside `idValidUrl`.

File: src/hooks/jukebox.ts [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+};
+
+const useConnectVenueJukeboxMessages = (venueId?: string, tableId?: string) => {
+  useFirestoreConnect(
+    venueId && tableId
+      ? {
+          collection: "venues",
+          doc: venueId,
+          subcollections: [{ collection: "jukeboxMessages" }],
+          orderBy: ["ts_utc", "desc"],
```
Comment (sunny-viktoryia) [inline src/hooks/jukebox.ts:109]: I've asked @sofisparkle. The history messages should be at the top.

**Review by dmitrymarchuk — APPROVED** [review]
Seems like all the comments from https://github.com/sparkletown/sparkle/pull/2209 are addressed. Thank you @alexey-petrov!

Approving besides a minor thing.

**Review by sunny-viktoryia — COMMENTED** [review]
first pass review from me. 🙃 

please, also fix the UI bugs I've sent you in slack.

File: src/components/templates/Jazzbar/JazzTab/JazzTab.tsx (hunk 7 above)
Comment (alexey-petrov) [reply src/components/templates/Jazzbar/JazzTab/JazzTab.tsx:135]: that's the necessary logic, so that only 'driver' table can update the iframe url, that's @sofisparkle 's requirement

Comment (sunny-viktoryia) [reply src/components/templates/Jazzbar/JazzTab/JazzTab.tsx:135]: I believe it's a BM requirement. But for the regular JazzBar venue, it shouldn't be like that?

Comment (sunny-viktoryia) [reply src/components/templates/Jazzbar/JazzTab/JazzTab.tsx:135]: and how is it going to work for BM? 
'driver' is at one table, and other people are at other tables? 
so If I change it at any table it changes for all the tables, correct?

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+            );
+          })}
+        </div>
+        <form className="Jukebox__form" onSubmit={sendMessageToChat}>
+          <InputField
+            containerClassName="Jukebox__input-container"
+            inputClassName="Jukebox__input"
+            ref={register({ required: true })}
+            name="jukeboxMessage"
+            placeholder="Add a link to the queue (YouTube, Vimeo, Twitch etc.)"
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx:121]: we process everything, but if it's youtube link I convert it to embed
image

File: src/types/jukebox.ts (hunk 1 above)
Comment (alexey-petrov) [reply src/types/jukebox.ts:7]: you can take seconds and convert it to miliseconds and then create new Date from it 
image

File: src/utils/jukebox.ts (hunk 3 above)
Comment (alexey-petrov) [reply src/utils/jukebox.ts:9]: ref

File: src/types/jukebox.ts (hunk 1 above)
Comment (sunny-viktoryia) [reply src/types/jukebox.ts:7]: not super convenient.. yeah, i saw that.

File: src/utils/url.ts (hunk 10 above)
Comment (alexey-petrov) [reply src/utils/url.ts]: fixed by removing new foo and refactoring the `isValidUrl` d04b6de

File: src/components/molecules/IframeAdmin/IframeAdmin.tsx [hunk 13]
```diff
@@ -20,18 +21,10 @@ export const IframeAdmin: React.FC<IframeAdminProps> = ({ venueId, venue }) => {
...
-      .functions()
-      .httpsCallable("venue-adminUpdateIframeUrl")({ venueId, iframeUrl })
-      .catch((e) => setError(e.toString()));
-  };
-
   const saveIframeUrl = (e: React.MouseEvent<HTMLButtonElement>) => {
     e.preventDefault();
 
-    updateIframeUrl(iframeUrl);
+    updateIframeUrl(iframeUrl, venueId).catch((err) => setError(err));
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/IframeAdmin/IframeAdmin.tsx]: fixed d04b6de

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+          {filteredMessages?.map((msg, index) => {
+            const msgAuthorName = msg.isMine
+              ? `${msg.author.partyName} (That's you))`
+              : msg.author.partyName;
+            return (
+              <span key={`${msg}${index}`} className="Jukebox__chat-messages">
+                {msgAuthorName} changed video source to {msg.text}
+              </span>
+            );
+          })}
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: done d04b6de

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+    <>
+      <div className="Jukebox__container">
+        <div className="Jukebox__chat">
+          <span className="Jukebox__chat-messages">
+            When you paste a link here, it changes what people see in the box
+            above. Be respectful!
+          </span>
+          {filteredMessages?.map((msg, index) => {
+            const msgAuthorName = msg.isMine
+              ? `${msg.author.partyName} (That's you))`
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: fixed d04b6de

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+              iconSize="1x"
+            />
+          </div>
+        </form>
+      </div>
+    </>
+  );
+};
+
+export default Jukebox;
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: fixed d04b6de

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+  venueName,
+}) => {
+  const { register, handleSubmit, watch, reset } = useForm<{
+    jukeboxMessage: string;
+  }>({
+    mode: "onSubmit",
+  });
+  const [isSendingMessage, setMessageSending] = useState(false);
+  const chatValue = watch("jukeboxMessage");
+  const venueId = useVenueId();
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: yep, did that d04b6de

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+  const { userId } = useUser();
+  const [filteredUser] =
+    recentVenueUsers.filter((vUser) => vUser.id === userId) || [];
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: fixed d04b6de

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+
+  const tableName = filteredUser?.data?.[venueName]?.table ?? "";
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: fixed d04b6de

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+
+  const { sendJukeboxMsg, messagesToDisplay } = useJukeboxChat({
+    venueId,
+    tableId: tableName,
+  });
+
+  const filteredMessages = messagesToDisplay.filter(
+    (msg) => msg.tableId === tableName
+  );
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: fixed d04b6de

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+      if (lastMessage?.text?.includes(YOUTUBE_SHORT_URL_STRING)) {
+        const youtubeUrl = getYoutubeEmbedFromUrl(lastMessage?.text);
+        updateIframeUrl(youtubeUrl);
+        return;
+      }
+      if (isStringAValidUrl(lastMessage?.text)) {
+        updateIframeUrl(lastMessage.text);
+        return;
+      }
+      return;
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: fixed d04b6de

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+            When you paste a link here, it changes what people see in the box
+            above. Be respectful!
+          </span>
+          {filteredMessages?.map((msg, index) => {
+            const msgAuthorName = msg.isMine
+              ? `${msg.author.partyName} (That's you))`
+              : msg.author.partyName;
+            return (
+              <span key={`${msg}${index}`} className="Jukebox__chat-messages">
+                {msgAuthorName} changed video source to {msg.text}
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: fixed d04b6de

File: src/components/templates/Jazzbar/JazzTab/JazzTab.scss [hunk 23]
```diff
@@ -118,6 +127,13 @@
           ".  video video ."
           ".  video video  ."
           "items items items items";
+
+        &-jukebox {
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/templates/Jazzbar/JazzTab/JazzTab.scss]: fixed and removed --jukebox modifier completely d04b6de

File: src/components/templates/Jazzbar/components/JazzBarRoom/Room.scss [hunk 24]
```diff
@@ -11,6 +11,10 @@
     max-height: $video-participant-optimal-height;
     border-radius: 15px;
 
+    &-jukebox {
+      max-height: $video-participant-jukebox-height;
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/templates/Jazzbar/components/JazzBarRoom/Room.scss]: removed that modifier d04b6de

File: src/components/templates/Jazzbar/components/JazzBarRoom/Room.tsx [hunk 25]
```diff
@@ -229,6 +232,10 @@ const Room: React.FC<RoomProps> = ({
     return [sidedVideoParticipants, otherVideoParticipants];
   }, [participants]);
 
+  const containerClasses = classNames("jazzbar-room__participant", {
+    "jazzbar-room__participant-jukebox": shouldShowJukebox,
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/templates/Jazzbar/components/JazzBarRoom/Room.tsx]: removed modifier d04b6de

File: src/components/templates/Jazzbar/jazzbar-constants.scss [hunk 26]
```diff
@@ -1,2 +1,3 @@
 $video-participant-optimal-width: 260px;
 $video-participant-optimal-height: 195px;
+$video-participant-jukebox-height: 300px;
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/templates/Jazzbar/jazzbar-constants.scss]: removed d04b6de

File: src/types/venues.ts (hunk 2 above)
Comment (alexey-petrov) [reply src/types/venues.ts]: done d04b6de

File: src/utils/url.ts (hunk 4 above)
Comment (alexey-petrov) [reply src/utils/url.ts]: refactored d04b6de

File: src/components/molecules/Jukebox/Jukebox.scss (hunk 5 above)
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.scss]: done d04b6de

File: src/components/templates/Jazzbar/JazzTab/JazzTab.tsx (hunk 8 above)
Comment (alexey-petrov) [reply src/components/templates/Jazzbar/JazzTab/JazzTab.tsx:192]: it's not clear to me what this comment is about? could you expand on it?

File: src/hooks/jukebox.ts (hunk 11 above)
Comment (alexey-petrov) [reply src/hooks/jukebox.ts]: done d04b6de

File: src/components/templates/Jazzbar/JazzTab/JazzTab.tsx (hunk 9 above)
Comment (alexey-petrov) [reply src/components/templates/Jazzbar/JazzTab/JazzTab.tsx]: removed f4b57f4

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+      }
+      return;
+    }
+  }, [messagesToDisplay, updateIframeUrl]);
+
+  const sendMessageToChat = handleSubmit(({ jukeboxMessage }) => {
+    setMessageSending(true);
+
+    sendJukeboxMsg({ message: jukeboxMessage });
+    reset();
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: fixed a00f7ff

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 28]
```diff
@@ -0,0 +1,142 @@
...
+    if (!isSendingMessage) return;
+
+    const timeoutId = setTimeout(() => {
+      setMessageSending(false);
+    }, CHAT_MESSAGE_TIMEOUT);
+
+    return () => {
+      clearTimeout(timeoutId);
+    };
+  }, [isSendingMessage]);
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx:92]: yep :)

File: src/components/templates/Jazzbar/components/JazzBarRoom/Room.scss [hunk 29]
```diff
@@ -11,6 +11,10 @@
     max-height: $video-participant-optimal-height;
     border-radius: 15px;
 
+    &-jukebox {
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/templates/Jazzbar/components/JazzBarRoom/Room.scss]: removed a00f7ff

File: src/settings.ts [hunk 30]
```diff
@@ -87,6 +87,8 @@ export const CURRENT_TIME_IN_LOCATION = "Matong State Forest";
 export const DUST_STORM_TEXT_1 = `A dust storm is ripping across the ${PLAYA_VENUE_NAME}!`;
 export const DUST_STORM_TEXT_2 =
   "Your only option is to seek shelter in a nearby venue!";
+export const YOUTUBE_SHORT_URL_STRING = "youtu";
+export const MUSIC_VENUE_FIRST_TABLE_NAME = "Table 1";
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/settings.ts]: removed a00f7ff & changed table find logic

File: src/components/molecules/TableHeader/TableHeader.scss [hunk 31]
```diff
@@ -23,6 +23,7 @@ $edit-topic-button-diameter: 0px 5px;
     padding: 15px 15px 15px 25px;
     border-radius: 0 0 16px 0;
     font-weight: $font-weight--500;
+    position: fixed;
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/TableHeader/TableHeader.scss]: fixed 9643d1c

File: src/api/jukebox.ts [hunk 32]
```diff
@@ -0,0 +1,34 @@
...
+
+export const sendJukeboxMessage = async ({
+  venueId,
+  message,
+  tableId,
+}: SendVenueMessageProps): Promise<void> =>
+  getVenueRef(venueId)
+    .collection("jukeboxMessages")
+    .add(message)
+    .then(noop)
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/api/jukebox.ts:22]: some issue with `.add` returning `void | | DocumentReference<DocumentData>` and not a promise, so we have to use then there

File: src/api/jukebox.ts [hunk 33]
```diff
@@ -0,0 +1,34 @@
...
+      console.log(err);
+      Bugsnag.notify(err, (event) => {
+        event.addMetadata("context", {
+          location: "api/chat::sendJukeboxMessage",
+          venueId,
+          message,
+          tableId,
+        });
+      });
+      // @debt rethrow error, when we can handle it to show UI error
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/api/jukebox.ts:32]: we don't seem to have any ui error handling here, so that's why I guess
but that was moved from chat api logic, so I don't know much about it

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 34]
```diff
@@ -0,0 +1,169 @@
...
+
+  return (
+    <>
+      <div className="Jukebox__container">
+        <div className="Jukebox__chat">
+          <span className="Jukebox__chat-messages">
+            JUKEBOX RULES: There’s no queue system. If you post up your link,
+            it’ll play. Be courteous people, do not post a new link until the
+            other one is finished!
+          </span>
```
Comment (sunny-viktoryia) [inline src/components/molecules/Jukebox/Jukebox.tsx:140]: let's give this text a bit opacity to distinguish it from users' messages. (opacity: 0.8 I guess should look good)

File: src/types/venues.ts [hunk 35]
```diff
@@ -245,6 +245,7 @@ export interface JazzbarVenue extends BaseVenue {
   host: {
     icon: string;
   };
+  enableJukebox?: boolean;
```
Comment (sunny-viktoryia) [inline src/types/venues.ts:248]: 😍  finally we start distinguishing properties between the venue types. mmmm. Thanks!

File: src/utils/url.ts [hunk 36]
```diff
@@ -92,13 +92,31 @@ export const openUrl = (url: string, options?: OpenUrlOptions) => {
   }
 };
 
-export const isValidUrl = (url: string): boolean => {
+export const isValidUrl = ({
```
Comment (sunny-viktoryia) [inline src/utils/url.ts:95]: let's refactor this! According to our discussion.

File: src/components/molecules/Jukebox/Jukebox.scss (hunk 6 above)
Comment (sunny-viktoryia) [reply src/components/molecules/Jukebox/Jukebox.scss:77]: is it possible to move to add a class name to the ButtonNG component? So that we wouldn't need to increase the specificity of that selector?

**Review by sunny-viktoryia — COMMENTED** [review]
few more comments. 🙃

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 37]
```diff
@@ -0,0 +1,169 @@
...
+  const { register, handleSubmit, watch, reset } = useForm<{
+    jukeboxMessage: string;
+  }>({
+    mode: "onSubmit",
+  });
+  const [isSendingMessage, setMessageSending] = useState(false);
+  const chatValue = watch("jukeboxMessage");
+  const { userId } = useUser();
+  const [filteredUser] = recentVenueUsers.filter(({ id }) => id === userId);
+  const tableRef = filteredUser?.data?.[venue.name]?.table ?? "";
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: fixed f0bbdc8

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 38]
```diff
@@ -0,0 +1,169 @@
...
+    let urlToEmbed = lastMessage?.text;
+    if (
+      urlToEmbed?.includes(YOUTUBE_SHORT_URL_STRING) &&
+      isValidUrl({ url: urlToEmbed, isStrict: true })
+    ) {
+      urlToEmbed = getYoutubeEmbedFromUrl(lastMessage?.text);
+    }
+    if (isValidUrl({ url: urlToEmbed, isStrict: true })) {
+      updateIframeUrl(urlToEmbed);
+    }
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: fixed f0bbdc8

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 39]
```diff
@@ -0,0 +1,169 @@
...
+            <span
+              className="Jukebox__chat-author"
+              onClick={() => openUserProfileModal(msg.author)}
+            >
+              {msg.author.partyName}{" "}
+            </span>
+            <span>{textMessage}</span>
+          </div>
+        );
+      }),
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: fixed f0bbdc8

File: src/components/templates/Jazzbar/JazzTab/JazzTab.tsx [hunk 40]
```diff
@@ -120,6 +127,13 @@ const Jazz: React.FC<JazzProps> = ({ setUserList, venue }) => {
 
   const shouldShowReactions =
     (seatedAtTable && venue.showReactions) ?? DEFAULT_SHOW_REACTIONS;
+  const firstTableReference = venue?.config?.tables?.[0].reference;
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/templates/Jazzbar/JazzTab/JazzTab.tsx]: fixed f0bbdc8

File: src/components/templates/Jazzbar/JazzTab/JazzTab.tsx [hunk 41]
```diff
@@ -168,25 +186,27 @@ const Jazz: React.FC<JazzProps> = ({ setUserList, venue }) => {
...
+                      frameBorder="0"
+                      allow={IFRAME_ALLOW}
+                    />
+                  ) : (
+                    <div className="iframe-video">
+                      Embedded Video URL not yet set up
+                    </div>
+                  )}
+                </div>
+              }
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/templates/Jazzbar/JazzTab/JazzTab.tsx]: fixed f0bbdc8

File: src/utils/url.ts (hunk 36 above)
Comment (alexey-petrov) [reply src/utils/url.ts]: fixed f0bbdc8

File: src/components/molecules/Jukebox/Jukebox.tsx (hunk 34 above)
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: fixed f0bbdc8

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 42]
```diff
@@ -0,0 +1,169 @@
...
+  const isBtnDisabled = !chatValue || isSendingMessage;
+  const submitButtonClasses = classNames("Jukebox__buttons-submit", {
+    "Jukebox__buttons-submit--active": !isBtnDisabled,
+  });
+
+  const messagesEndRef = useRef<null | HTMLDivElement>(null);
+
+  const scrollToBottom = () => {
+    messagesEndRef.current?.scrollIntoView({ behavior: "smooth" });
+  };
```
[the remark this answers is not in the document]
Comment (alexey-petrov) [reply src/components/molecules/Jukebox/Jukebox.tsx]: fixed f0bbdc8

File: src/components/molecules/Jukebox/Jukebox.tsx [hunk 43]
```diff
@@ -0,0 +1,171 @@
...
+  const { openUserProfileModal } = useProfileModalControls();
+
+  const jukeboxChatMessages = useMemo(
+    () =>
+      filteredMessages?.map((msg) => {
+        const isUrl = isValidUrl(msg.text);
+        return (
+          <div key={msg.id} className="Jukebox__chat-messages">
+            <span
+              className="Jukebox__chat-author"
```
Comment (sunny-viktoryia) [inline src/components/molecules/Jukebox/Jukebox.tsx:119]: let's make them more obvious to click. This class name makes them look like URLs adding yellow color on hover.
```suggestion
              className="Jukebox__chat-author button--a"
```

**Review by sunny-viktoryia — APPROVED** [review]
Wooo!! Looks good! Thank you!! 😍 
One tiny comment left.
